map.go

```diff
@@ -579,6 +581,133 @@ func (m *Map) nextKey(key interface{}, nextKeyOut internal.Pointer) error {
 	return nil
 }
 
+// BatchLookup looks up many elements in a map at once
+// with the startKey being the first element to start
+// from.
+func (m *Map) BatchLookup(startKey, nextKey, keysOut, valuesOut interface{}, opts *BatchOptions) (int, error) {
+	return m.batchLookup(false, startKey, nextKey, keysOut, valuesOut, opts)
+}
+
+// BatchLookup looks up many elements in a map at once
+// with the startKey being the first element to start
+// from. It then deletes all those elements.
+func (m *Map) BatchLookupAndDelete(startKey, nextKey, keysOut, valuesOut interface{}, opts *BatchOptions) (int, error) {
+	return m.batchLookup(true, startKey, nextKey, keysOut, valuesOut, opts)
+}
+
+func (m *Map) batchLookup(del bool, startKey, nextKey, keysOut, valuesOut interface{}, opts *BatchOptions) (int, error) {
+	if err := haveBatchAPI(); err != nil {
+		return 0, err
+	}
+	if !m.typ.canBatch() || (del && !m.typ.canBatchDelete()) {
+		return 0, ErrBatchOpNotSup
+	}
+	keysValue := reflect.ValueOf(keysOut)
+	if keysValue.Kind() != reflect.Slice {
+		return 0, fmt.Errorf("keys must be a slice")
+	}
+	valuesValue := reflect.ValueOf(valuesOut)
+	if valuesValue.Kind() != reflect.Slice {
+		return 0, fmt.Errorf("valuesOut must be a slice")
+	}
+	count := keysValue.Len()
+	if count != valuesValue.Len() {
+		return 0, fmt.Errorf("keysOut and valuesOut must be the same length")
+	}
+	keyBuf := make([]byte, count*int(m.keySize))
+	keyPtr := internal.NewSlicePointer(keyBuf)
+	valueBuf := make([]byte, count*int(m.valueSize))
+	valuePtr := internal.NewSlicePointer(valueBuf)
+
+	var (
+		startPtr internal.Pointer
+		err      error
+	)
+	if startKey != nil {
+		startPtr, err = marshalPtr(startKey, int(m.keySize))
+		if err != nil {
+			return 0, err
+		}
+	}
+
+	nextPtr, nextBuf := makeBuffer(nextKey, int(m.keySize))
+	var ct uint32
+	if del {
+		ct, err = bpfMapBatch(internal.BPF_MAP_LOOKUP_AND_DELETE_BATCH, m.fd, startPtr, nextPtr, keyPtr, valuePtr, uint32(count), opts)
+	} else {
+		ct, err = bpfMapBatch(internal.BPF_MAP_LOOKUP_BATCH, m.fd, startPtr, nextPtr, keyPtr, valuePtr, uint32(count), opts)
+	}
+	if err != nil && !errors.Is(err, ErrKeyNotExist) {
```

Comment:
In which case is ErrKeyNotExist returned? When startKey doesn't exist?

Reply:
When it reaches the end of the batch lookups, but it doesn't happen for arrays.

Reply:
The fact that we don't get ENOENT for arrays seems really screwed up! Looking at https://elixir.bootlin.com/linux/latest/source/kernel/bpf/syscall.c#L1364 makes me think that a non existing key on an array will return EINTR, which seems mad. I think that this can be triggered by doing a batch lookup on an array where the batch has num > max_entries or starts above max_entries.

Maybe you can provide me some context here: how is BatchLookup supposed to be used? It seems to me like this would be called in a loop, which bails out once there are no more keys. It seems sensible to return ErrKeyNotExist in that case, wdyt? If you agree, what do we do about arrays that return EINTR?

Reply:
Yeah, arrays needs some special consideration, but I think it makes sense that they don't behave the same as maps. There's no real way of knowing to what degree an array is populated and if a "key" (actually ordinal) is "there" or not.

Reply:
I think arrays should behave just the same as hashmaps. If the batch is larger than the array we return `count = len(array)` and `err = ErrNotExist`.

Reply:
> I think arrays should behave just the same as hashmaps.

In a sense they do. If you think of arrays like they're hash maps then you can think of arrays as being pre-populated, so you'll never get the ENOENT until you hit the end of the whole array. I do like the idea of checking the count of the keys compared to the size of the map, but that's probably all we can do. We can't get smart about things, like taking `startKey` into account for example. Does that make sense?

Reply:
> If you think of arrays like they're hash maps then you can think of arrays as being pre-populated, so you'll never get the ENOENT until you hit the end of the whole array.

Yes, with the wrinkle that due to a (IMO) implementation bug arrays never return ENOENT, they return EINTR. This is probably something we should report to bpf@

> I do like the idea of checking the count of the keys compared to the size of the map, but that's probably all we can do.

My idea boils down to: `if map type == array && errno == EINTR return cnt, ErrKeyNotExist`. This works because arrays don't have a legitimate reason for EINTR, since keys can't be deleted in mid iteration like for hash maps.

What do you mean by checking the count of the keys?

Reply:
I'm not seeing any error actually, even when the full array is looked up. When keys/values beyond the size of the array are requested I'm seeing ENOENT.